### DISCUSSION TITLE
Can I configure a key sequence that freeRDP passes through to the local operating system?

### DISCUSSION THREAD

**Discussion by bernd-wechner** [FreeRDP/FreeRDP#9273] [Q&A]
I like having grab-keyboard on, so that on my work desktop ALT+TAB as expected.

But similarly, I'd like a particular set of keystrokes to pass through. I would like for example for the keypress:

`<ALT>+<CRL>+<RIGHT>`

and 

`<ALT>+<CRL>+<LEFT>`

to pass through to the operating system (as these move between local workspaces).

But to be honest, it would be nice to be able to flexible define key sequences that are passed through to the local OS when freeRDP is grabbing the keyboard.  Of conversely of course, if -grab-keyboard is specified that one could specify a list of defined key sequences that it grabs (like ALT+TAB) - though I suspect that is a none sequitur (not possible, as the whole point of grabbing the keyboard is to see those sequences before the OS grabs them and freeRDP never sees them).

**Comment by akallabeth** [comment]
that depends on which client you use ;)

as for `xfreerdp` you can define a `ActionScript` with key combinations that will call the script with the combination as argument.

**Reply by bernd-wechner** [reply]
Yes, xfreerdp (wasn't aware there were others, but I guess there are clients for different window managers?). Alas I am not sure what "call the script with the combination as argument." means. I'd like to capture those keys and pass them through to the underlying desktop environment (to be processed locally).

**Reply by bernd-wechner** [reply]
In fact the docs on this are near non existant!

https://www.mankier.com/1/xfreerdp#Keyboard_Shortcuts

is the best I can find. And so I put this:

```
#!/bin/bash

# Define the log file
LOG_FILE="$HOME/.config/freerdp/keystrokes.log"

# Log all arguments with a timestamp
echo "$(date '+%Y-%m-%d %H:%M:%S') - Arguments: $@" >> "$LOG_FILE"
```

in `$HOME/.config/freerdp/action.sh` and ran rdp and hits some funky keys. Nothing. So I ran rdp with `/action-script:$HOME/.config/freerdp/action.sh` and I do get something logged (great docs, they say that's the default file, but it's ignored unless you use the command line option, oh well). 

But nowhere does it say when this script is called. It can't possibly call it for every keystork that would be a massive performance hit. So for what keystokes? It says "The script will receive the current key combination as argument. " yet all this logs after many keystrokes is:

```
2024-07-24 21:14:38 - Arguments: key
2024-07-24 21:14:38 - Arguments: xevent
```

So no, key combination is not passed, only the work "key" and it's not at all clear when it's called, what triggers calling it in order to not pass the keystrokes to it. Shrug.

**Reply by akallabeth** [reply]
@bernd-wechner happy about pull requests ;)
I´ve updated the whole thing in https://github.com/FreeRDP/FreeRDP/pull/10423 but feedback and testing is very welcome.

**Reply by dollhaus** [reply]
@bernd-wechner did you ever find out what keystrokes trigger the action script?

**Reply by bernd-wechner** [reply]
Nope ☹️

**Reply by dollhaus** [reply]
Hm. I tailed the keyboard log as taken from your test and discovered that the script is called both on "key" and "xevent", but that those two are fired before the rdp session starts (and then seemingly never again). 

There's something to do with key-local here, whcih I will test now

Edit: Nope, nothing. I did note that the usual hotkeys (ctrl-alt-enter etcetera) don't actually appear in the log, which is odd. 

Edit2: Interesting. Adding the argument `-grab-keyboard` makes the key and xevent events fire twice. Still nothing in the actual session though

Final edit: I've given up. I've resorted to specifying `-grab-keyboard` and then using a kwin-managed script that uses xdotool to accomplish what I wanted. Not ideal, but it works for me for the moment.

**Comment by SamSirry** [comment]
I think there's another perfectly valid use case for keys like `Monitor Brightness Control`, `Sleep`, `Hibernate`, and the `Power Button` that should be allowed to be processed locally instead of getting blocked/redirected.
